### ISSUE TITLE
Integration Test Scans break on 6th scan

### ISSUE THREAD

**Issue by T1960CT** [boskokg/flutter_blue_plus#94]
Hello, I'm writing an integration test for a bluetooth device management app, and any time I try to load the app (pumpWidget) for the 6th time, the scan results are always empty. The FlutterBluePlus.instance.startScan() is done in initState for 3 seconds, but I've also reworked a branch of my original project to only start scanning on floatingActionButton press, but same issues.

I had the same issue with espresso's fork which led me to this one, so I'm not sure if it's an issue with the original flutter_blue plugin that made it here, or if there's something wrong with either how I'm approaching this or the phone itself.

Added in build.gradle
```gradle
android {
  defaultConfig {
    minSdkVersion 19
    targetSdkVersion 31
    compileSdkVersion 31
  }
}
```

Added in AndroidManifest.xml
```xml
<!-- USER ADDED -->
    <uses-permission android:name="android.permission.BLUETOOTH" android:maxSdkVersion="30"/>
    <uses-permission android:name="android.permission.BLUETOOTH_ADMIN" android:maxSdkVersion="30"/>
    <uses-permission android:name="android.permission.BLUETOOTH_SCAN"/>
    <uses-permission android:name="android.permission.BLUETOOTH_CONNECT" />
    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />

    <uses-feature android:name="android.hardware.bluetooth"/>
    <uses-feature android:name="android.hardware.bluetooth_le"/>
```

flutter doctor
```shell
[✓] Flutter (Channel stable, 3.0.1, on macOS 12.2.1 21D62 darwin-arm, locale en-CA)
[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
[✗] Xcode - develop for iOS and macOS
    ✗ Xcode installation is incomplete; a full installation is necessary for iOS development.
      Download at: https://developer.apple.com/xcode/download/
      Or install Xcode via the App Store.
      Once installed, run:
        sudo xcode-select --switch /Applications/Xcode.app/Contents/Developer
        sudo xcodebuild -runFirstLaunch
    ✗ CocoaPods not installed.
        CocoaPods is used to retrieve the iOS and macOS platform sides plugin code that responds to your
        plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To install see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.
[✓] Chrome - develop for the web
[✓] Android Studio (version 2021.1)
[✓] VS Code (version 1.69.2)
[✓] Connected device (3 available)
[✓] HTTP Host Availability
```
Phone is Pixel 6 with latest Android 12 update July 5 (SQ3A.220705.003),
I only run Android apps for the time being, so I don't think I need the Xcode dependencies, unless that's part of the issue.

Basically, bluetooth functions just fine and I've had no issues sending or receiving data, scanning, service and characteristic discovery, etc. Just seems to lock up on the 6th app launch.

I made a repo with minimum viable reproduction. I commented out the 5th and 6th test in the first group to show it makes it to the second group before hitting the 6th total test, if you un comment those lines, it fails to reach the second group.

Can someone take a look and let me know if I'm making an obvious mistake or if there really is a problem?
https://gitlab.com/T1960CT/scantest

Thanks in advance!

**Comment by SaikCaskey** [comment]
Don't have a link or anything but there's something documented somewhere about this - pretty sure android itself stops you scanning so many times within a brief period. I mean actually you shouldn't need to scan so much, scanning is expensive and unless you're doing real-time triangulation or something, you should mitigate your accesses to the platform resources. If you do need high sample updates I think there's a newer permission for that and probably some justification to make in the play store. 

To test this, you could make a start/stop scan button to simulate what happens during your app launch, probably around the 6th time you'll see no scan results  - I do see this myself and testers were kind of annoyed that you can't just start and stop scans all day, so we looked about and found that you're not really intended to do that - but to be honest, it makes good sense, the point of scanning is not to start and stop scans; usually the point is to find a device to connect to, and in most cases, a single scan is completely fine, especially in dart where you get a nice stream structure to listen to to process each set of results as it comes in

**Comment by T1960CT** [comment]
Of course, magically after you say this I find a link referencing exactly what you're saying, even though I spent hours scouring before asking 😅 🙄 always seems to go this way, haha. There's a limitation of 5 BLE scans per 30 seconds since Android 7, according to [this](https://proandroiddev.com/android-restrictions-you-may-encounter-during-development-process-c39ede513813#:~:text=Bluetooth%20scan%20limitations&text=Starting%20from%20Android%207%20you,will%20work%20in%20a%20loop.) link.

For this particular application, I definitely don't expect an end user to need to make that many scans, just doing so in integration tests. For my test setup, I have one bluetooth dev board simulating a proper unit, and I bounce data back and forth between them. From what I now understand, I should run one scan when needed, and store the results to refer back to in subsequent tests? The issue I was having when trying to plan this out is getting a BluetoothDevice to connect to in each test, but since it's the same device and in a controlled environment (IE not wandering range, I know it's always nearby), there'd be nothing wrong with holding onto the BluetoothDevice I get, and assuming it's there when I connect?

Thanks for cluing me in, it was SO consistent it felt like it had to be system level!

**Comment by SaikCaskey** [comment]
Awesome to hear @T1960CT, you described the problem well, I'm glad it wasn't (necessarily) a problem! And thanks for this link by the way, it's super useful to communicate about this! 👌
